File: python_stack/learn_assignments/bank_account/bank_account.py
# 
# 

#  Bank Account:

class BankAccount:
    def __init__(self, int_rate, balance):
        self.int_rate = int_rate 
        self.balance = 0

    def deposit(self, amount):
        self.balance += amount
        print("Deposit: ", self.balance)
        return self

    def withdrawal(self, amount):
        if self.balance > amount:
            self.balance -= amount
        if self.balance < amount:
            print("Insufficient Funds: Charging a $5 fee.")
            self.balance -= amount +5
            print("Withdrawal: ", self.balance)
        return self

    def display_account_info(self):
        print("Balance: ", self.balance)
        return self

    def yield_int(self):
        self.balance += self.balance * self.int_rate
        print("Interest: ")
        return self

    def transfer_money(self, other_user, amount):  
        other_user.account_balance += amount
        print("Transfer: ", self.balance)
        return self

File: java_2/lectures/OOP_1/BankAccount.java
public class BankAccount {
    public String firstName;
    public String lastName;
    private Double accountBalance;


    // This is my constructor (similar to def __init__() method in python.)

    public BankAccount(String firstNameInput, String lastNameInput) {
        this.firstName = firstNameInput;
        this.lastName = lastNameInput;
        this.accountBalance = 0.00;
    }

    // overloading a constructor. 
    
    // This creates a bank account without needing any parameters.

    public BankAccount() {
        this.firstName = "untitled firstname";
        this.lastName = "untitled lastname";
        this.accountBalance = 0.00;
    }

    // this creates and account and also displays the initial amount. Three constructors.

    public BankAccount(String firstNameInput, String lastNameInput, Double initialAmount) {
        this.firstName = firstNameInput;
        this.lastName = lastNameInput;
        this.accountBalance = initialAmount;
    }

    // overloading a method.

    public Double deposit(Double amount) {
        this.accountBalance += amount;
        return this.accountBalance;
    }

    public Double deposit() {
        this.accountBalance += 10;
    }


// A getter to get the account balance. A getter allows you to access a variable.

    public Double getAccountBalance() {
        return this.accountBalance;
    }

    // a setter to modify the account balance. (deposit add to the balance. Setters are void and do not return anything.) Setters allow us to modify private variables.

    public void setAccountBalance(Double newBalance) {
        this.accountBalance = newBalance;
    }

    // who to send money to. (which account)
    // the amount you want to send. 
    public void transferMoney(BankAccount otherAccount, Double amountToSend) {
        // the object that called the transferMoney method is refferred to with the keyword "this".
        this.accountBalance-= amountToSend;
        otherAccount.accountBalance += amountToSend;
    }

    public String displayAccountInfo() {
        // return "Account Balance: {}, First Name: {}, Last Name: {}"
            // return "Account Balanace: " + this.accountBalance + "First Name: " + this.firstName + "Last Name" + this.lastName;

            return String.format("AccountBalance: %f, First Name: %s , Last Name: {}", this.accountBalance, this.firstName, this.lastName);
    }

}